File: src/main/java/com/sems/analytics/interfaces/rest/AnalyticsController.java
package com.sems.analytics.interfaces.rest;

import com.sems.analytics.application.AnalyticsService;
import com.sems.analytics.interfaces.rest.resources.AnalyticsResources.*;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

/**
 * API REST del bounded context de analitica.
 *
 * <p>Rutas identicas a las del microservicio en FastAPI, bajo el prefijo
 * {@code /api/v1/analytics}. El frontend consume sobre todo
 * {@code /bill-predictions/user/{id}} y {@code /consumption-rankings/user/{id}}.
 */
@Tag(name = "Analytics", description = "Proyecciones, recomendaciones, anomalias y rankings")
@RestController
@RequestMapping("/api/v1/analytics")
@RequiredArgsConstructor
public class AnalyticsController {

    private final AnalyticsService service;

    // -------------------------------------------------------- bill-predictions

    @Operation(summary = "Proyecciones de recibo de un usuario")
    @GetMapping("/bill-predictions/user/{userId}")
    public List<BillPredictionResponse> predictions(@PathVariable String userId) {
        return service.predictionsByUser(userId).stream().map(BillPredictionResponse::from).toList();
    }

    @Operation(summary = "Registra una proyeccion de recibo")
    @PostMapping("/bill-predictions")
    public ResponseEntity<BillPredictionResponse> createPrediction(
            @Valid @RequestBody CreatePredictionRequest r) {
        var created = service.createPrediction(r.userId(), r.predictionYear(), r.predictionMonth(),
                r.periodStart(), r.periodEnd(), r.estimatedKwh(), r.estimatedAmount(),
                r.currency(), r.tariffUsed(), r.errorMarginPercentage());
        return ResponseEntity.status(HttpStatus.CREATED).body(BillPredictionResponse.from(created));
    }

    // ---------------------------------------------------------- recommendations

    @Operation(summary = "Recomendaciones de ahorro de un usuario")
    @GetMapping("/recommendations/user/{userId}")
    public List<RecommendationResponse> recommendations(@PathVariable String userId) {
        return service.recommendationsByUser(userId).stream()
                .map(RecommendationResponse::from).toList();
    }

    @Operation(summary = "Registra una recomendacion")
    @PostMapping("/recommendations")
    public ResponseEntity<RecommendationResponse> createRecommendation(
            @Valid @RequestBody CreateRecommendationRequest r) {
        var created = service.createRecommendation(r.userId(), r.deviceId(), r.recommendationType(),
                r.title(), r.description(), r.estimatedSavingKwh(), r.estimatedSavingAmount(),
                r.currency());
        return ResponseEntity.status(HttpStatus.CREATED).body(RecommendationResponse.from(created));
    }

    @Operation(summary = "Marca una recomendacion como aplicada")
    @PatchMapping("/recommendations/{recommendationId}/apply")
    public RecommendationResponse applyRecommendation(@PathVariable UUID recommendationId) {
        return RecommendationResponse.from(service.applyRecommendation(recommendationId));
    }

    // ---------------------------------------------------------------- anomalies

    @Operation(summary = "Anomalias detectadas para un usuario")
    @GetMapping("/anomalies/user/{userId}")
    public List<AnomalyResponse> anomalies(@PathVariable String userId) {
        return service.anomaliesByUser(userId).stream().map(AnomalyResponse::from).toList();
    }

    @Operation(summary = "Registra una anomalia")
    @PostMapping("/anomalies")
    public ResponseEntity<AnomalyResponse> createAnomaly(@Valid @RequestBody CreateAnomalyRequest r) {
        var created = service.createAnomaly(r.userId(), r.deviceId(), r.anomalyType(),
                r.description(), r.severity(), r.actualKwh(), r.expectedKwh());
        return ResponseEntity.status(HttpStatus.CREATED).body(AnomalyResponse.from(created));
    }

    @Operation(summary = "Da por resuelta una anomalia")
    @PatchMapping("/anomalies/{anomalyId}/resolve")
    public AnomalyResponse resolveAnomaly(@PathVariable UUID anomalyId) {
        return AnomalyResponse.from(service.resolveAnomaly(anomalyId));
    }

    // ------------------------------------------------------ device-identifications

    @Operation(summary = "Identificaciones de aparatos de un usuario")
    @GetMapping("/device-identifications/user/{userId}")
    public List<DeviceIdentificationResponse> identifications(@PathVariable String userId) {
        return service.identificationsByUser(userId).stream()
                .map(DeviceIdentificationResponse::from).toList();
    }

    @Operation(summary = "Registra una identificacion de aparato")
    @PostMapping("/device-identifications")
    public ResponseEntity<DeviceIdentificationResponse> createIdentification(
            @Valid @RequestBody CreateIdentificationRequest r) {
        var created = service.createIdentification(r.userId(), r.deviceId(),
                r.predictedDeviceType(), r.confidenceScore(), r.status());
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(DeviceIdentificationResponse.from(created));
    }

    // ------------------------------------------------------- consumption-rankings

    @Operation(summary = "Rankings de consumo de un usuario")
    @GetMapping("/consumption-rankings/user/{userId}")
    public List<ConsumptionRankingResponse> rankings(@PathVariable String userId) {
        return service.rankingsByUser(userId).stream().map(ConsumptionRankingResponse::from).toList();
    }

    @Operation(summary = "Registra un ranking de consumo")
    @PostMapping("/consumption-rankings")
    public ResponseEntity<ConsumptionRankingResponse> createRanking(
            @Valid @RequestBody CreateRankingRequest r) {
        var items = r.rankings() == null ? List.<com.sems.analytics.domain.model.valueobjects.RankingItem>of()
                : r.rankings().stream().map(RankingItemResource::toDomain).toList();
        var created = service.createRanking(r.userId(), r.periodType(), r.periodStart(),
                r.periodEnd(), items);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ConsumptionRankingResponse.from(created));
    }
}

File: src/Sems.Api/Modules/Analytics/Interfaces/AnalyticsController.cs
using Microsoft.AspNetCore.Mvc;
using Sems.Api.Modules.Analytics.Application;
using Sems.Api.Modules.Analytics.Domain.Model;
using static Sems.Api.Modules.Analytics.Interfaces.AnalyticsResources;

namespace Sems.Api.Modules.Analytics.Interfaces;

/// <summary>
/// API REST del bounded context de analitica.
///
/// <para>Rutas identicas a las del microservicio en FastAPI, bajo el prefijo
/// <c>/api/v1/analytics</c>. El frontend consume sobre todo
/// <c>/bill-predictions/user/{id}</c> y <c>/consumption-rankings/user/{id}</c>.</para>
/// </summary>
[ApiController]
[Route("api/v1/analytics")]
[Tags("Analytics")]
public sealed class AnalyticsController : ControllerBase
{
    private readonly AnalyticsService _service;

    public AnalyticsController(AnalyticsService service) => _service = service;

    // ---------------------------------------------------- bill-predictions

    /// <summary>Proyecciones de recibo de un usuario.</summary>
    [HttpGet("bill-predictions/user/{userId}")]
    public async Task<List<BillPredictionResponse>> Predictions(string userId) =>
        (await _service.PredictionsByUserAsync(userId)).Select(BillPredictionResponse.From).ToList();

    /// <summary>Registra una proyeccion de recibo.</summary>
    [HttpPost("bill-predictions")]
    public async Task<ActionResult<BillPredictionResponse>> CreatePrediction(
        [FromBody] CreatePredictionRequest r)
    {
        var created = await _service.CreatePredictionAsync(r.UserId, r.PredictionYear,
            r.PredictionMonth, r.PeriodStart, r.PeriodEnd, r.EstimatedKwh, r.EstimatedAmount,
            r.Currency, r.TariffUsed, r.ErrorMarginPercentage);
        return StatusCode(StatusCodes.Status201Created, BillPredictionResponse.From(created));
    }

    /// <summary>
    /// Calcula la proyeccion de recibo de un local con la tarifa comercial,
    /// incluido el cargo por potencia.
    /// </summary>
    [HttpPost("bill-predictions/forecast")]
    public async Task<ActionResult<BillPredictionResponse>> ForecastSiteBill(
        [FromBody] ForecastSiteBillRequest r)
    {
        var created = await _service.ForecastSiteBillAsync(r.UserId, r.SiteId, r.TariffCategory,
            r.ContractedPowerKw, r.PredictionYear, r.PredictionMonth, r.PeriodStart, r.PeriodEnd,
            r.KwhPeak, r.KwhOffPeak, r.MaxDemandKw, r.ErrorMarginPercentage);
        return StatusCode(StatusCodes.Status201Created, BillPredictionResponse.From(created));
    }

    // ------------------------------------------------------ recommendations

    /// <summary>Recomendaciones de ahorro de un usuario.</summary>
    [HttpGet("recommendations/user/{userId}")]
    public async Task<List<RecommendationResponse>> Recommendations(string userId) =>
        (await _service.RecommendationsByUserAsync(userId)).Select(RecommendationResponse.From).ToList();

    /// <summary>Registra una recomendacion.</summary>
    [HttpPost("recommendations")]
    public async Task<ActionResult<RecommendationResponse>> CreateRecommendation(
        [FromBody] CreateRecommendationRequest r)
    {
        var created = await _service.CreateRecommendationAsync(r.UserId, r.DeviceId,
            r.RecommendationType, r.Title, r.Description, r.EstimatedSavingKwh,
            r.EstimatedSavingAmount, r.Currency);
        return StatusCode(StatusCodes.Status201Created, RecommendationResponse.From(created));
    }

    /// <summary>Marca una recomendacion como aplicada.</summary>
    [HttpPatch("recommendations/{recommendationId:guid}/apply")]
    public async Task<RecommendationResponse> ApplyRecommendation(Guid recommendationId) =>
        RecommendationResponse.From(await _service.ApplyRecommendationAsync(recommendationId));

    // ------------------------------------------------------------ anomalies

    /// <summary>Anomalias detectadas para un usuario.</summary>
    [HttpGet("anomalies/user/{userId}")]
    public async Task<List<AnomalyResponse>> Anomalies(string userId) =>
        (await _service.AnomaliesByUserAsync(userId)).Select(AnomalyResponse.From).ToList();

    /// <summary>Registra una anomalia.</summary>
    [HttpPost("anomalies")]
    public async Task<ActionResult<AnomalyResponse>> CreateAnomaly([FromBody] CreateAnomalyRequest r)
    {
        var created = await _service.CreateAnomalyAsync(r.UserId, r.DeviceId, r.AnomalyType,
            r.Description, r.Severity, r.ActualKwh, r.ExpectedKwh);
        return StatusCode(StatusCodes.Status201Created, AnomalyResponse.From(created));
    }

    /// <summary>Da por resuelta una anomalia.</summary>
    [HttpPatch("anomalies/{anomalyId:guid}/resolve")]
    public async Task<AnomalyResponse> ResolveAnomaly(Guid anomalyId) =>
        AnomalyResponse.From(await _service.ResolveAnomalyAsync(anomalyId));

    // ------------------------------------------- device-identifications

    /// <summary>Identificaciones de aparatos de un usuario.</summary>
    [HttpGet("device-identifications/user/{userId}")]
    public async Task<List<DeviceIdentificationResponse>> Identifications(string userId) =>
        (await _service.IdentificationsByUserAsync(userId))
        .Select(DeviceIdentificationResponse.From).ToList();

    /// <summary>Registra una identificacion de aparato.</summary>
    [HttpPost("device-identifications")]
    public async Task<ActionResult<DeviceIdentificationResponse>> CreateIdentification(
        [FromBody] CreateIdentificationRequest r)
    {
        var created = await _service.CreateIdentificationAsync(r.UserId, r.DeviceId,
            r.PredictedDeviceType, r.ConfidenceScore, r.Status);
        return StatusCode(StatusCodes.Status201Created, DeviceIdentificationResponse.From(created));
    }

    // -------------------------------------------- consumption-rankings

    /// <summary>Rankings de consumo de un usuario.</summary>
    [HttpGet("consumption-rankings/user/{userId}")]
    public async Task<List<ConsumptionRankingResponse>> Rankings(string userId) =>
        (await _service.RankingsByUserAsync(userId)).Select(ConsumptionRankingResponse.From).ToList();

    /// <summary>Registra un ranking de consumo.</summary>
    [HttpPost("consumption-rankings")]
    public async Task<ActionResult<ConsumptionRankingResponse>> CreateRanking(
        [FromBody] CreateRankingRequest r)
    {
        var items = r.Rankings?.Select(i => i.ToDomain()) ?? Enumerable.Empty<RankingItem>();
        var created = await _service.CreateRankingAsync(r.UserId, r.PeriodType, r.PeriodStart,
            r.PeriodEnd, items);
        return StatusCode(StatusCodes.Status201Created, ConsumptionRankingResponse.From(created));
    }
}
